crates/nu-cli/src/types/deduction.rs

```diff
@@ -0,0 +1,1451 @@
+                                    fake_bin.left.expr = l_fake_var_expr;
+                                    let op = op_of(source_bin);
+                                    self.dependencies_on_result_type.push((
+                                        var.clone(),
+                                        op,
+                                        SpannedExpression::new(
+                                            Expression::Binary(fake_bin),
+                                            source_bin.span,
+                                        ),
+                                    ));
+                                    None
+                                }
+                            }
+                        }
+                        //After here every invocation on get_shape_of_expr_or_insert_dependency(expr) should
+                        //give a result shape
+                        (Expression::Binary(_), _) => {
+                            let (l_fake_var, l_fake_var_expr) =
+                                self.fake_var_generator.next_as_expr();
+                            let (r_fake_var, _) = self.fake_var_generator.next_as_expr();
+                            let fake_bin = change_op_to_assignment(expr.clone());
+                            match (
+                                self.get_shape_of_binary_arg_or_insert_dependency(
+                                    (&l_fake_var, &bin.left),
+                                    &fake_bin,
+                                    (pipeline_idx, pipeline),
+                                    registry,
+                                )?,
+                                self.get_shape_of_expr_or_insert_dependency(
+                                    &r_fake_var,
+                                    (Operator::Equal, source_bin.span),
+                                    &bin.right,
+                                ),
+                            ) {
+                                (Some(l_shape), Some(r_shape)) => {
+                                    Some(get_result_shape_of(l_shape, &bin.op, r_shape)?)
+                                }
+                                (None, _) => {
+                                    self.substitute_left_with_fake_var_and_insert_dependencies(
+                                        bin,
+                                        &l_fake_var_expr,
+                                        (source_bin, var),
+                                    );
+                                    None
+                                }
+                                (Some(_), None) => {
+                                    unreachable!("At this point shape should be deducable!")
+                                }
+                            }
+                        }
+                        (_, Expression::Binary(_)) => {
+                            let (l_fake_var, _) = self.fake_var_generator.next_as_expr();
```

Comment:
With this fake var pieces, I'm wondering if what we need is instead a flattening phase. That way, we don't need to embed it into the inference code, but instead can run a pass over the AST that flattens all the nested expressions and assigns the simple expressions to unique variables.

Like:

```
$var * 1 + 1
```
from above would be replaced with:

```
$unique_var_12345 = $var * 1
$unique_var_12346 = $unique_var_12345 + 1
```

This type of expansion is pretty common with compilers, and I believe if we do it correctly should help simplify the code here and also help us iron out any corner cases which might arise.

Reply:
Because of your input, I have removed this piece of code. Instead of creating fake_vars and building dependencies for all sub-binaries, lets push one dependency of a variable to the other side of the binary. 
That makes the code also easier to grasp and will result in the same thing.